repart: add new settings Format=, CopyFiles=, Encrypt= and teach --size= a new value "auto"

poettering opened the pull request.

mrc0mmand commented:
There appears to be a memory leak triggered by `test-repart`:

```
Successfully formatted future partition 6.
Populating partition 6 with files.
Successfully populated partition 6 with files.
Registering future partition 6 in partition table.
Writing new partition table.
All done.

=================================================================
==48389==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 66 byte(s) in 6 object(s) allocated from:
    #0 0x7f1368b44d51 in strdup (/usr/lib/clang/10.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xb4d51)
    #1 0x55ddc1b7f1eb in context_load_partition_table /systemd-meson-build/../build/src/partition/repart.c:1576:38
    #2 0x55ddc1b793aa in run /systemd-meson-build/../build/src/partition/repart.c:4026:13
    #3 0x55ddc1b7900a in main /systemd-meson-build/../build/src/partition/repart.c:4122:1
    #4 0x7f136756f001 in __libc_start_main (/usr/lib/libc.so.6+0x27001)

SUMMARY: AddressSanitizer: 66 byte(s) leaked in 6 allocation(s).
```

See: https://ci.centos.org/job/systemd-pr-build-vagrant-sanitizers/7173/artifact//systemd-centos-ci/artifacts_all/artifacts_Pr4aK8/vagrant-logs.8Oy/vagrant-arch-sanitizers-clang-testsuite.9ox/ninja-test_sanitizers_FAIL.log

poettering commented:
@mrc0mmand indeed. the bug is actually older, the new test triggers it for the first time now... i added another commit now, that fixes it. let's see if this passes.

behrmann commented on man/repart.d.xml:
```diff
@@ -327,7 +327,71 @@
+      <varlistentry>
+        <term><varname>Format=</varname></term>
+
+        <listitem><para>Takes a file system name, such as <literal>ext4</literal>, <literal>btrfs</literal>,
+        <literal>xfs</literal> or <literal>vfat</literal>, or the special value <literal>swap</literal>. If
+        specified and the partition is newly created it is formatted with the specified file system (or as
+        swap device). The file system UUID and label are automatically derived from the partition UUID and
+        label. If this option is used, the size allocation algorithm is slightly altered: the partition is
```
If the file system UUID is derived from the partition UUID, which would be a small pool considering the discoverable partition scheme, and the label, which would usually be something guessable, wouldn't there be a high chance for file system UUIDs coming out the same? I vaguely remember this being a potential issue for btrfs when they two file systems with the same UUID are mounted on the same machine. Could this be a problem?
I'm sorry if this is a silly question.

poettering replied:
GPT manages two UUIDs per partition. The partition UUID and the partition type UUID. The former is supposed to identify the partition uniquely world-wide, and is what we derive from the seed that is either the machine ID or random or whatever you specify. The latter identifies the type of the partition and is what is defined in the discoverable partition spec.

the gpt partition uuid should hence be as unique as that seed value you specify, and the since the fs uuid is derived from the partition uuid, the same is true for that.

Does that make sense?

behrmann replied:
That makes sense. Thank you for the explanation!

keszybz commented on src/partition/repart.c:
```diff
@@ -2691,7 +2633,7 @@ static int context_write_partition_table(
                                 return log_error_errno(r, "Failed to set partition label: %m");
 
-                        log_info("Creating new partition %" PRIu64 ".", p->partno);
+                        log_info("Registering new partition %" PRIu64 " in partition table.", p->partno);
```
"Adding new partition ... to partition table" would be even nicer I think.

keszybz commented on src/partition/repart.c:
```diff
@@ -2702,6 +2644,76 @@ static int context_write_partition_table(
+        assert(context);
+
+        if (arg_pretty > 0 ||
+            (arg_pretty < 0 && isatty(STDOUT_FILENO) > 0) || arg_json) {
```
`|| arg_json` deserves it's own line.

keszybz commented on src/partition/repart.c:
```diff
@@ -2702,6 +2644,76 @@ static int context_write_partition_table(
+        if (from_scratch) {
+                r = context_discard_range(context, 0, context->total);
+                if (r == -EOPNOTSUPP)
+                        log_info("Storage does not support discarding, not discarding entire block device data.");
```
"does not support discard" (we call the capability "discard" everywhere).

keszybz commented on src/partition/repart.c:
```diff
@@ -2110,17 +2110,17 @@ static int context_discard_partition(Context *context, Partition *p) {
         }
         if (r < 0)
-                return log_error_errno(r, "Failed to discard data for new partition %" PRIu64 ".", p->partno);
+                return log_error_errno(r, "Failed to discard data for future partition %" PRIu64 ".", p->partno);
```
Yep, "future partition" is much clearer.

keszybz commented on src/partition/repart.c:
```diff
@@ -1035,22 +1041,95 @@ static int config_parse_fstype(
+                return 0;
+        }
+
+        r = extract_first_word(&p, &buffer, "", EXTRACT_CUNESCAPE|EXTRACT_DONT_COALESCE_SEPARATORS);
```
With no separators, this is essentially `cunescape`. Wouldn't it be better to just call cunescape directly?

keszybz commented on src/partition/repart.c:
```diff
@@ -371,8 +379,25 @@ static uint64_t partition_min_size(const Partition *p) {
+                else if (p->format || p->encrypt) {
+                        uint64_t f;
+
+                        /* If we shall synthesize a file system, take minimal fs size into account (assumed to be 4K if not known) */
```
I guess this calculation will not be enough, since it's unlikely that 4k is enough for most filesystems. But I guess it's a good start. If people find cases where this fails, we can easily add the fs to minimal_size_by_fs_name().

keszybz commented on man/repart.d.xml:
```diff
@@ -327,7 +327,71 @@
+        newly created partition and formatted file system. This setting may be used to copy files or
+        directories from the host into the file system that is created due to the <varname>Format=</varname>
+        option. If <varname>CopyFiles=</varname> is used without <varname>Format=</varname> specified
+        explicitly, <literal>Format=ext4</literal> is implied. This option may be used multiple times to copy
```
Maybe let's make this more flexible: "Format= with a suitable default is implied (currently 'ext4', but this may change in the future)".

keszybz commented on man/repart.d.xml:
```diff
@@ -327,7 +327,71 @@
+        option. If <varname>CopyFiles=</varname> is used without <varname>Format=</varname> specified
+        explicitly, <literal>Format=ext4</literal> is implied. This option may be used multiple times to copy
+        multiple files or directories from host into the newly formatted file system. The colon and second
+        path may be omitted in which case the target path is chosen identical to the source path — however
```
"in which case the source path is also used as the target path (relative to the root of the newly created file system)" ?

keszybz commented on man/systemd-repart.xml:
```diff
@@ -202,13 +202,26 @@
+        <para>The <option>--size=auto</option> option takes the sizes of pre-existing partitions into
+        account. However, it does not accomodate for partition tables that are not tightly packed: the
+        configured partitions might still not fit into the backing device if empty space exists between
+        pre-existing partitions (or before the first partition) that cannot be fully filled be partitions to
```
be → by

keszybz reviewed (commented):
This is super cool: this is now turning into a swiss-army-knife system installer thingy.
Some minor comments and suggestions.

poettering commented:
Force pushed a new version: made the changes @keszybz suggested. nothing else — except for the final commit, which i added anew. I couldn't get the PR pass the tests anymore. Turns out the new LOOP_CONFIGURE ioctl on kernels 5.8 is even more broken than previously determine: the .lo_sizelimit parameter is not applied properly by LOOP_CONFIGURE. Hence I added another check: right after issuing LOOP_CONFIGURE we'll check if the block device size matches our expectations, and if not we'll fallback to the old ways.

Kinda sucks that the new ioctl — that could be so useful — is so broken, it's really as if more stuff is broken on it than works. Given that the earlier bugfix I sent to the kernel (for the partition scan loop device issue) is apparently not going to be backported onto 5.8 as it looks now (though merged for 5.9) I think it probably makes sense to carry this new, additional work-around for a while. I mean, normally I think we should be agressive with requiring non-broken dependencies in cases like this, but this is harder to expect for the kernel, in particular where it appears to be nothing people intend to backport.

poettering commented:
(for those interested I started a thread on lkml and related mls regarding this new LOOP_CONFIGURE bug. not sure i want to fix this one upstream myself, too though: https://lore.kernel.org/linux-block/20200824154436.GA255257@gardel-login/T/#u)

poettering commented:
anyway, dropping green label given the new workaround commit that wasn#t there before. please have a look!

keszybz commented on src/partition/repart.c:
```diff
@@ -1030,22 +1036,95 @@ static int config_parse_fstype(
+        }
+
+        r = extract_first_word(&p, &buffer, ":", EXTRACT_CUNESCAPE|EXTRACT_DONT_COALESCE_SEPARATORS);
+        if (r < 0)
```
Yeah, that makes more sense now.

keszybz commented on src/partition/repart.c:
```diff
@@ -2307,16 +2315,139 @@ static int context_wipe_and_discard(Context *context, bool from_scratch) {
+
+        assert(node);
+
+        /* udev or so might access out block device in the background while we are done. Let's hence force
```
This first sentence doesn't make much sense: "out", "while we are done"...

keszybz commented on man/repart.d.xml:
```diff
@@ -55,11 +55,11 @@
-    <citerefentry><refentrytitle>systemd-growfs</refentrytitle><manvolnum>8</manvolnum></citerefentry> and
-    <command>systemd-makefs</command> maybe be used to initialize or grow the file systems inside of these
-    partitions.</para>
+    <para>Note that these definitions may only be used to created and initialize new partitions or grow
```
To "create"

keszybz commented on man/repart.d.xml:
```diff
@@ -55,11 +55,11 @@
-    <command>systemd-makefs</command> maybe be used to initialize or grow the file systems inside of these
-    partitions.</para>
+    <para>Note that these definitions may only be used to created and initialize new partitions or grow
+    existing ones. In the latter case it will not grow the contained files systems however; separate
```
"file systems"

keszybz reviewed (commented):
OK, let's merge this.
